### PR TITLE
feat(hostpath): enforce quotas for hostpath with an ext4 filesystem

### LINKED ISSUE #136: Enforce ext4 quotas
**Describe the problem/challenge you have**
The setup I'm using has an EXT4 filesystem and switching to XFS is not a viable option.  Adding quotas for ext4 filesystems would solve that problem.

**Describe the solution you'd like**
In similar fashion to https://github.com/openebs/dynamic-localpv-provisioner/pull/78, the update will enable EXT4 project quotas to be enforced.

**Anything else you would like to add:**
This expands ext4 support for issue: https://github.com/openebs/dynamic-localpv-provisioner/issues/13
I've updated the code locally with tests and I'll perform a pull request for review. 

**Environment:**
- OpenEBS version (use `kubectl get po -n openebs --show-labels`): `3.2.0`
- Kubernetes version (use `kubectl version`): `v1.24.2`
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): `Fedora 33`
- kernel (e.g: `uname -a`): `Linux endor 5.10.99 #23 SMP Tue Jun 28 12:29:23 PDT 2022 x86_64 x86_64 x86_64 GNU/Linux`
- others:

### PR DESCRIPTION
**Pull request by hickersonj**
Signed-off-by: Jacoby Hickerson jacoby.hickerson@outlook.com

**Why is this PR required? What issue does it fix?**:
Quotas have been implemented on xfs filesystems, this PR extends that to ext4 filesystems. 
This PR fixes issue: https://github.com/openebs/dynamic-localpv-provisioner/issues/136

**What this PR does?**:
This PR enforces quotas on the hostpath that is formatted with an ext4 filesystem.

**Does this PR require any upgrade changes?**:
No

**If the changes in this PR are manually verified, list down the scenarios covered:**:
The setup, deployment and test instructions can be found here:
https://pastebin.com/PK6XVDkp

**Any additional information for your reviewer?** : 
This PR is a contribution to the xfs quota enforcement pr: https://github.com/openebs/dynamic-localpv-provisioner/pull/78

**Checklist:**
- [ ] Fixes #136
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated? 
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track: 
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

**Comment by hickersonj** [comment]
I've created a pr: https://github.com/openebs/linux-utils/pull/26 for the inclusion of e2fsprogs and quota-tools in the dependent linux-tools image, which will be needed for some of these tests to pass.

Additionally I was unaware of updating the chart version, and I'm not sure what version it should be bumped to.

File: tests/hostpath_quota_test.go [hunk 1]
```diff
@@ -82,15 +82,15 @@ var _ = Describe("TEST HOSTPATH XFS QUOTA LOCAL PV WITH NON-XFS FILESYSTEM", fun
...
+			By("Writing minix filesystem into the loop device")
 			//umount
 			errs := physicalDisk.Unmount()
 			Expect(errs).To(BeEmpty(), "when unmounting loop device")
 			//wipefs
 			err := physicalDisk.Wipefs()
 			Expect(err).To(BeNil(), "when wiping filesystem from loop device")
 			//mkfs
-			err = physicalDisk.CreateFilesystem("ext4")
+			err = physicalDisk.CreateFilesystem("minix")
```
Comment (niladrih) [inline tests/hostpath_quota_test.go:93]: Same change is required in line 199.

File: tests/hostpath_quota_test.go [hunk 2]
```diff
@@ -407,3 +407,382 @@ var _ = Describe("TEST HOSTPATH XFS QUOTA LOCAL PV WITH XFS FILESYSTEM", func()
...
+
+			By("verifying PVC is deleted")
+			status := ops.IsPVCDeletedEventually(pvcName, namespaceObj.Name)
+			Expect(status).To(
+				BeTrue(),
+				"when checking status of deleted PVC {%s}",
+				pvcName,
+			)
+
+			By("removing the xfs filesystem")
```
Comment (niladrih) [inline tests/hostpath_quota_test.go]: 'xfs' should be 'minix'

**Review by niladrih — CHANGES_REQUESTED** [review]
Thank you for your contribution @hickersonj !

The code changes look good, the helm chart changes however are handled in a separate PR which is merged when we create a new release. 

Please pull in the `helm-3.3.0` branch and push your helm chart commits to that branch.
They will show up in this PR, and this will be merged when we create a new release (15th July).
- https://github.com/openebs/dynamic-localpv-provisioner/pull/139

File: tests/hostpath_quota_test.go (hunk 2 above)
Comment (hickersonj) [reply tests/hostpath_quota_test.go]: Good eye thanks!

**Comment by hickersonj** [comment]
@niladrih thank you for reviewing the changes!  
I pushed the helm chart commits to my forked tree but I don't think it quite did what we wanted.  Let me know the step I missed and I can try again.

```
# from within my forked project
git fetch upstream helm-3.3.0
git checkout helm-3.3.0
git rebase upstream/helm-3.3.0

# made my changes, added and committed

git push origin helm-3.3.0
Enumerating objects: 17, done.
Counting objects: 100% (17/17), done.
Delta compression using up to 8 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (9/9), 1.03 KiB | 1.03 MiB/s, done.
Total 9 (delta 5), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (5/5), completed with 5 local objects.
remote:
remote: Create a pull request for 'helm-3.3.0' on GitHub by visiting:
remote:      https://github.com/hickersonj/dynamic-localpv-provisioner/pull/new/helm-3.3.0
remote:
To https://github.com/hickersonj/dynamic-localpv-provisioner.git
 * [new branch]      helm-3.3.0 -> helm-3.3.0
```
Can find the changes here:
https://github.com/hickersonj/dynamic-localpv-provisioner/tree/helm-3.3.0

**Comment by niladrih** [comment]
No issue @hickersonj , I'll rearrange the commits and sort it out. Thanks :)

**Review by niladrih — APPROVED** [review]

File: cmd/provisioner-localpv/app/helper_hostpath.go [hunk 3]
```diff
@@ -279,18 +280,24 @@ func (p *Provisioner) createQuotaPod(ctx context.Context, pOpts *HelperPodOption
...
+	//PID is the last project Id in the directory
+	//xfs_quota project(xfs) or chattr +P (ext4) initializes project with new project id
+	//xfs_quota limit(xfs) or repquota (ext4) sets the quota according to limits defined
+	checkQuota := "" +
+		"if [[ \"$FS\" == \"xfs\" ]]; then " +
+		"  PID=`xfs_quota -x -c 'report -h' /data | tail -2 | awk 'NR==1{print substr ($1,2)}+0'` ;" +
+		"  PID=`expr $PID + 1` ;" +
+		"  xfs_quota -x -c 'project -s -p " + filepath.Join("/data/", config.volumeDir) + " '$PID /data;" +
+		"  xfs_quota -x -c 'limit -p bsoft=" + config.pOpts.softLimitGrace + " bhard=" + config.pOpts.hardLimitGrace + " '$PID /data ;" +
+		"elif [[ \"$FS\" == \"ext2/ext3\" ]]; then" +
```
Comment (Ab-hishek) [inline cmd/provisioner-localpv/app/helper_hostpath.go:293]: What about ext4?

Comment (niladrih) [reply cmd/provisioner-localpv/app/helper_hostpath.go:293]: The stat output says `ext2/ext3` in ext4 filesystem.

**Review by Ab-hishek — APPROVED** [review]
LGTM.

File: cmd/provisioner-localpv/app/helper_hostpath.go (hunk 3 above)
Comment (Ab-hishek) [reply cmd/provisioner-localpv/app/helper_hostpath.go:293]: Ohkk. Got it.

Comment (hickersonj) [reply cmd/provisioner-localpv/app/helper_hostpath.go:293]: Yes that's correct, even though the FS is ext4, the output of `stat -f -c %T /mount` is `ext2/ext3`.  
Initially I was thinking I would also add a md similar to https://github.com/openebs/dynamic-localpv-provisioner/blob/develop/docs/tutorials/hostpath/xfs_quota/enable-xfs-quota.md with some prerequisites, but I wasn't sure if that would be too many changes for this PR.  Also I wasn't sure if someone else writes the documentation, didn't want to step on any toes.

**Pull request merged**